### ISSUE TITLE
Coercion error complains about sentinel when const is the problem

### ISSUE THREAD

**Issue by SpexGuy** [ziglang/zig#7816]
Tested with 0.7.1 and godbolt's master.
```zig
export fn b() void {
    var ptr: [*:0]u8 = "foo";
}
```
```
<source>:2:24: error: expected type '[*:0]u8', found '*const [3:0]u8'
    var ptr: [*:0]u8 = "foo";
                       ^
<source>:2:24: note: destination pointer requires a terminating '0' sentinel
    var ptr: [*:0]u8 = "foo";
                       ^
```
This complains about the sentinel, but both types have the correct sentinel.  Instead, the const mismatch is what is causing the problem.  This also happens on coercion of return expressions.

**Comment by LemonBoy** [comment]
> This complains about the sentinel, but both types have the correct sentinel.

Not so fast. `*const [3:0]u8` as a pointer has no sentinel, it's the child type, the array, to have the same sentinel.

**Comment by SpexGuy** [comment]
Yes, what I mean is that the sentinel is not preventing coercion.  As far as it's concerned, the sentinels in the two types are compatible.
